src/gmtset.1
.TH GMTSET 1 "Feb 27 2014" "GMT 4.5.13 (SVN)" "Generic Mapping Tools"
.SH NAME
gmtset \- To change individual \fBGMT\fP default parameters
.SH SYNOPSIS
\fBgmtset\fP [ \fB\-G\fP\fIdefaultsfile\fP ] PARAMETER1 [=] \fIvalue1\fP PARAMETER2 [=] \fIvalue2\fP PARAMETER3 [=] \fIvalue3\fP ...
.SH DESCRIPTION
\fBgmtset\fP will adjust individual \fBGMT\fP defaults settings in the current directory's \.gmtdefaults4\"'
file.  If no such file exists one will be created.  The main purpose of \fBgmtset\fP is
temporarily to change certain parameters inside a shell script, e.g., set the dots-per-inch
to 600, run the script, and reset to 300 dpi.  Optionally, you can specify one or more temporary changes
directly on any \fBGMT\fP command line with the syntax \fB\-\-PARAMETER\fP=\fIvalue\fP; such changes are
only in effect for that command and do not permanently change the default settings on disk.
.TP
PARAMETER \fIvalue\fP
Provide one or several pairs of parameter/value combinations that you want to modify.  For a
complete listing of available parameters and their meaning, see the \fBgmtdefaults\fP man page.
.SH OPTIONS
.TP
\fB\-G\fP
Name of specific \.gmtdefaults4 file to modify [Default looks first in current directory, then in your
home directory, then in ~/.gmt and finally in the system defaults].
.SH EXAMPLES
To change the dpi to 600, set annotation font to Helvetica, and select grid-crosses of size 0.1 inch,
and set annotation offset to 0.2 cm:
.br
.sp
\fBgmtset\fP \fBDOTS_PR_INCH\fP 600 \fBANNOT_FONT_PRIMARY\fP Helvetica \fBGRID_CROSS_SIZE_PRIMARY\fP 0.1\fBi\fP \fBANNOT_OFFSET_PRIMARY\fP 0.2\fBc\fP
.SH "SEE ALSO"
.IR GMT (1),
.IR gmtdefaults (1),
.IR gmtget (1)


ChangeLog
-----------------------------------------------------------------------------------------
	$Id: ChangeLog 12925 2014-02-20 22:02:24Z pwessel $
-----------------------------------------------------------------------------------------

-----------------------------------------------------------------------------------------
201x-yy-zz	pwessel
	GMT 5.1.1 finalized and released.
-----------------------------------------------------------------------------------------
2014-02-20	pwessel
	* grdcontour.c:		The -Zscale/offset option was applied incorrectly due to a
				change in the underlying scaling function.
	* grdvolume.c:		---"---
2014-02-10	pwessel
	*gmt2kml.c:		Make sure the default style IDs are unique for each process.

2014-02-03	pwessel
	* pslib.c:		Auto-widen paragraph width to at least fit widest word.

2014-01-26	pwessel
	* gmt_grdio.c:		Building the grid command string could exceed its length if
				the argument became exactly the max length of the array; then
				a final extra character \0 would be added and exceed array length.
				Thanks to Joachim Saul for pointing it out.

2014-01-22	pwessel
	* pscoast.c:		The processing of -I -N levels and pens could get out of sync
				so that the wrong pen was used for the specified feature.
2014-01-16	remko
	* pslib.c:		Take superscripts and subscripts into account when determining
				dimensions of text box.

2014-01-10	pwessel
	* grdmath.c:		We consider the change in SDIST, PDIST to return degrees instead
				of km to be a bug as it makes it difficult to handle geodesics
				for which degrees make no sense.  Thus we have consolidated all
				the ?DIST operators to give km if geographic data are involved.
				Example 23 has been adjusted accordingly, as has several test
				scripts.  Please check your own scripts for SDIST.

2014-01-09	pwessel
	* gmt_vector.c:		Add protection against introduced longitude-jumps produced
				by resampling in GMT_fix_up_path.
2014-01-09	remko
	* gmt_vector.c:		Add protection against introduced longitude-jumps produced
				by resampling in GMT_fix_up_path.
	* grdmath.c:		LDIST and PDIST will again produce km when using geodesics.

2014-01-01	pwessel
	* x2sys/x2sys.c:	Mishandled the assignment of segment number for each record.
	* gmt_support.c:	Allow GMT_crossover to deal with global wrapped data.

2013-12-31	pwessel
	* gmt2kml.c:		Now using GMT_Put_Record instead of old stdout printing.
				This addresses issue #473.

2013-12-10	pwessel
	* grdvector.c:		Vector direction did not adjust when negative scales were
				used in -JX or -Jx.
	* psxy[z].c		---"---

2013-12-08	remko
	* psimage.c:		Now properly plots transparent images and makes -Gt work.

2013-11-26	pwessel
	* gmt_io.c:		GMT would crash on a blank line with no newline at end at all.
	* pstext.c:		pstext did not process GMT4-style input properly.

2013-11-26	jluis
	* potential/gravfft.c:	Let -E options default to 3.
				Let users select the -N+l (leave trend alone). That is, do not overwrite with default mode.
				Let the default detrend be remove MID point.
				Explicitly set -N+a in gravfft command to avoid failures due to coming changes in gravfft.
2013-11-25	pwessel
	* grdlandmask.c:	Failed to fill partial bins with no coastlines when the region
				crossed Greenwich [issue #451].

2013-11-21	pwessel
	* pscontour.c:		Let pscontour output contours to file [issue #448].
	* pscoast.c:		Prevent users from not giving proper projection [issue #449].

2013-11-19	jluis
	* mgd77/mgd77magref.c:	Needed to reset the t_col variable when constant altitude [issue #446].

2013-11-19	pwessel
	* gmt_sph.c:		Fix invalid read in ssrfpack_grid due to par array being of length 2 rather than 3.

2013-11-18	pwessel
	* gmt_grdio.c:		Place the old auto-detect of 4-byte binary float vs int under compatibility mode for backwardness.

2013-11-16	pwessel
	* grdvector.c:		Address shortcomings of geovectors when wrapped and clipped.

2013-11-15	pwessel
	* geovectors.c:		Fix geovector when shrinking is selected.

2013-11-13	pwessel
	* gmt_api.c:		Fix API issues pointed out by Ian.

2013-11-13	fwobbe
	* GMT.app:		Inherit user shell in GMT.app.

2013-11-12	pwessel
	* gmt_init.c:		grdmath could corrupt the gmt.history file [issue # 438].
 	
2013-11-10	pwessel
	* gmt_support.c:	Fix symbol-size issue in psxy -Sc [issue #435].
	
2013-11-09	pwessel
	* grdvector.c:		Did not convert direction to azimuths for geo vectors.

2013-11-08	pwessel
	* grdvector.c:		Fix grdvector -I problem [issue #434].
	* ps2raster.c:		Under Mavericks, strcpy would exit if the to and from pointers
-----------------------------------------------------------------------------------------
2013-11-05	pwessel
	GMT 5.1.0 finalized and released.
-----------------------------------------------------------------------------------------

2013-11-01	remko
	* example_10.sh:	Changed input to number of languages; multi-color columns

2013-10-31	pwessel
	* gmt_dcw.c:		Added option to specify code as =<continent> which will
				automatically be expanded to all the countries in that
				continent.  E.g., -F=AF+gred will paint Africa read.
				Fixed memory leaks.

2013-10-23	pwessel
	* gmtinfo.c:		The new name of minmax.c, the latter will be honored
				when GMT compatibility mode is 4.
	* grdmath.c:		Fixed porting bugs for CURV and D2DY2; added test ops.sh

2013-10-22	pwessel
	* minmax.c:		Added -D to adjust region found with -I
	* makecpt.c:		Better reporting of gaps or overlaps in CPTs

2013-10-05	pwessel
	* greenspline.c:	Needed -f in order to select -f0T input, plus it made assumptions
				about lon,lat despite not being selected.

2013-10-04	pwessel
	* gmt.history:		This replaces the old .gmtcommands file name.

2013-09-28	remko
	* gmt_init.c:		Moved origin of JD by one day (24 Nov instead of 25 Nov).
	* ConfigUserTemplate.cmake:	Added option GMT_DOCS_DEPEND_ON_GMT

2013-09-27	pwessel
	* share/cpt:		Added 10 new CPTs, mostly for bathymetry and topography.

2013-09-14	pwessel

src/gmtset.c
/*--------------------------------------------------------------------
 *	$Id: gmtset.c 12822 2014-01-31 23:39:56Z remko $
 *
 *	Copyright (c) 1991-2014 by P. Wessel, W. H. F. Smith, R. Scharroo, J. Luis and F. Wobbe
 *	See LICENSE.TXT file for copying and redistribution conditions.
 *
 *	This program is free software; you can redistribute it and/or modify
 *	it under the terms of the GNU Lesser General Public License as published by
 *	the Free Software Foundation; version 3 or any later version.
 *
 *	This program is distributed in the hope that it will be useful,
 *	but WITHOUT ANY WARRANTY; without even the implied warranty of
 *	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *	GNU Lesser General Public License for more details.
 *
 *	Contact info: gmt.soest.hawaii.edu
 *--------------------------------------------------------------------*/
/*
 * Author:	Paul Wessel
 * Date:	1-JAN-2010
 * Version:	5 API
 *
 * Brief synopsis: gmtset will override specified defaults parameters.
 *
 */
 
#define THIS_MODULE_NAME	"gmtset"
#define THIS_MODULE_LIB		"core"
#define THIS_MODULE_PURPOSE	"Change individual GMT default parameters"

#include "gmt_dev.h"

#define GMT_PROG_OPTIONS "-V" GMT_SHORTHAND_OPTIONS

/* Control structure for gmtset */

struct GMTSET_CTRL {
	struct C {	/* -C */
		bool active;
	} C;
	struct D {	/* -D[s|u] */
		bool active;
		char mode;
	} D;
	struct G {	/* -Gfilename */
		bool active;
		char *file;
	} G;
};

void *New_gmtset_Ctrl (struct GMT_CTRL *GMT) {	/* Allocate and initialize a new control structure */
	struct GMTSET_CTRL *C;

	C = GMT_memory (GMT, NULL, 1, struct GMTSET_CTRL);
	return (C);
}

void Free_gmtset_Ctrl (struct GMT_CTRL *GMT, struct GMTSET_CTRL *C) {	/* Deallocate control structure */
	if (!C) return;
	if (C->G.file) free (C->G.file);
	GMT_free (GMT, C);	
}

int GMT_gmtset_usage (struct GMTAPI_CTRL *API, int level)
{
	GMT_show_name_and_purpose (API, THIS_MODULE_LIB, THIS_MODULE_NAME, THIS_MODULE_PURPOSE);
	if (level == GMT_MODULE_PURPOSE) return (GMT_NOERROR);
	GMT_Message (API, GMT_TIME_NONE, "usage: gmtset [-C | -D[s|u] | -G<defaultsfile>] [-[" GMT_SHORTHAND_OPTIONS "]<value>] PARAMETER1 [=] value1 PARAMETER2 [=] value2 PARAMETER3 [=] value3 ...\n");
	GMT_Message (API, GMT_TIME_NONE, "\n\tFor available PARAMETERS, see gmt.conf man page.\n");

	if (level == GMT_SYNOPSIS) return (EXIT_FAILURE);

	GMT_Message (API, GMT_TIME_NONE, "\n\tOPTIONS:\n");
	GMT_Message (API, GMT_TIME_NONE, "\t-C Convert GMT4 .gmtdefaults4 to GMT5 gmt.conf file.\n");
	GMT_Message (API, GMT_TIME_NONE, "\t   The original file is retained.\n");
	GMT_Message (API, GMT_TIME_NONE, "\t-D Modify the default settings based on the GMT system defaults.\n");
	GMT_Message (API, GMT_TIME_NONE, "\t   Append s to see the SI version of defaults.\n");
	GMT_Message (API, GMT_TIME_NONE, "\t   Append u to see the US version of defaults.\n");
	GMT_Message (API, GMT_TIME_NONE, "\t-G Set name of specific gmt.conf file to modify.\n");
	GMT_Message (API, GMT_TIME_NONE, "\t   [Default looks for file in current directory.  If not found,\n");
	GMT_Message (API, GMT_TIME_NONE, "\t   it looks in the home directory, if not found it uses GMT defaults.]\n");
	GMT_Message (API, GMT_TIME_NONE, "\n\tThe modified defaults are written to the current directory as gmt.conf.\n");
	GMT_Message (API, GMT_TIME_NONE, "\n\t-[" GMT_SHORTHAND_OPTIONS "]<value> (any of these options).\n");
	GMT_Message (API, GMT_TIME_NONE, "\t   Set the expansion of any of these shorthand options.\n");
	
	return (EXIT_FAILURE);
}

int GMT_gmtset_parse (struct GMT_CTRL *GMT, struct GMTSET_CTRL *Ctrl, struct GMT_OPTION *options)
{
	/* This parses the options provided to gmtset and sets parameters in CTRL.
	 * Any GMT common options will override values set previously by other commands.
	 * It also replaces any file names specified as input or output with the data ID
	 * returned when registering these sources/destinations with the API.
	 */

	unsigned int n_errors = 0;
	struct GMT_OPTION *opt = NULL;

	for (opt = options; opt; opt = opt->next) {
		switch (opt->option) {

			case '<':	/* Input files: Here, arguments like ANNOT_OFFSET = -0.1i will show up as "files" or numbers */
			case '#':
				break;

			/* Processes program-specific parameters */

			case 'C':	/* Convert GMT4 .gmtdefaults4 to GMT5 gmt.conf */
				Ctrl->C.active = true;
				break;
			case 'D':	/* Get GMT system-wide defaults settings */
				Ctrl->D.active = true;
				Ctrl->D.mode = opt->arg[0];
				break;
			case 'G':	/* Optional defaults file on input and output */
				if ((Ctrl->G.active = GMT_check_filearg (GMT, 'G', opt->arg, GMT_IN)))
					Ctrl->G.file = strdup (opt->arg);
				else
					n_errors++;
				break;

			default:	/* Report bad options */
				n_errors += GMT_default_error (GMT, opt->option);
				break;
		}
	}
	
	return (n_errors ? GMT_PARSE_ERROR : GMT_OK);
}

#define bailout(code) {GMT_Free_Options (mode); return (code);}
#define Return(code) {Free_gmtset_Ctrl (GMT, Ctrl); GMT_end_module (GMT, GMT_cpy); bailout (code);}

int GMT_gmtset (void *V_API, int mode, void *args)
{
	int error = 0;

	char path[GMT_LEN256] = {""};
	char* gmtconf_file;

	struct GMTSET_CTRL *Ctrl = NULL;
	struct GMT_CTRL *GMT = NULL, *GMT_cpy = NULL;
	struct GMT_OPTION *options = NULL;
	struct GMTAPI_CTRL *API = GMT_get_API_ptr (V_API);	/* Cast from void to GMTAPI_CTRL pointer */

	/*----------------------- Standard module initialization and parsing ----------------------*/

	if (API == NULL) return (GMT_NOT_A_SESSION);
	if (mode == GMT_MODULE_PURPOSE) return (GMT_gmtset_usage (API, GMT_MODULE_PURPOSE));	/* Return the purpose of program */
	options = GMT_Create_Options (API, mode, args);	if (API->error) return (API->error);	/* Set or get option list */

	if (options) {
		if (options->option == GMT_OPT_USAGE) bailout (GMT_gmtset_usage (API, GMT_USAGE));		/* Return the usage message */
		if (options->option == GMT_OPT_SYNOPSIS) bailout (GMT_gmtset_usage (API, GMT_SYNOPSIS));	/* Return the synopsis */
	}

	/* Parse the command-line arguments */

	GMT = GMT_begin_module (API, THIS_MODULE_LIB, THIS_MODULE_NAME, &GMT_cpy); /* Save current state */
	if (GMT_Parse_Common (API, GMT_PROG_OPTIONS, options)) Return (API->error);
	Ctrl = New_gmtset_Ctrl (GMT);	/* Allocate and initialize a new control structure */
	if ((error = GMT_gmtset_parse (GMT, Ctrl, options))) Return (error);

	/*---------------------------- This is the gmtset main code ----------------------------*/

	/* Read the supplied default file or the users defaults to override system settings */

	if (Ctrl->D.active) {
		switch (Ctrl->D.mode) {
			case 's':
				gmtconf_file = "gmt_SI.conf";
				break;
			case 'u':
				gmtconf_file = "gmt_US.conf";
				break;
			default:
				gmtconf_file = "gmt.conf";
				break;
		}

		if (! GMT_getsharepath (GMT, "conf", "", gmtconf_file, path, R_OK))
			GMT_Report (API, GMT_MSG_NORMAL, "Cannot find GMT configuration file: %s (%s)\n", gmtconf_file, path);
		GMT_getdefaults (GMT, path);
	}
	else if (Ctrl->C.active)
		GMT_getdefaults (GMT, ".gmtdefaults4");
	else if (Ctrl->G.active)
		GMT_getdefaults (GMT, Ctrl->G.file);

	if (GMT_setdefaults (GMT, options)) Return (GMT_PARSE_ERROR);		/* Process command line arguments, return error if failures */

	GMT_putdefaults (GMT, Ctrl->G.file);	/* Write out the revised settings */

	Return (GMT_OK);
}
